### PR TITLE
Bump to v2.0.8: Bootstrap VMWatch Process if vmWatchSettings present

### PR DESCRIPTION
**Pull request by frank-pang-msft**
## Overview
This PR contains changes to support running VMWatch (amd64 and arm64) as an executable via goroutines and channels. 
> VMWatch is a standardized, lightweight, and open-sourced testing framework designed to enhance the monitoring and management of guest VMs on the Azure platform, including both 1P and 3P instances. VMWatch is engineered to collect vital health signals across multiple dimensions, which will be seamlessly integrated into Azure's quality systems. By leveraging these signals, VMWatch will enable Azure to swiftly detect and prevent regressions induced by platform updates or configuration changes, identify gaps in platform telemetry, and ultimately improve the guest experience for all Azure customers.

## Behavior
VMWatch will run asynchronously as a separate process than ApplicationHealth, so the probing of application health will not be affected by the state of VMWatch. Depending on extension settings, VMWatch can be enabled/disabled, and also specify the test names and parameter overrides to VMWatch binary. The status of VMWatch will be displayed in the extension x.status files and also in GET VM Instance View. Main process will attempt to start VMWatch binary up to 3 times, after which VMWatch status will be set to failed.

## Process Leaks
To ensure that VMWatch processes do not accumulate, applicationhealth-shim will be responsible for killing existing VMWatch processes by looking for processes running with the VMWatch binary names according to the architecture type. For unexpected process termination, if for some reason the main applicationhealth-extension process is terminated, we also ensure that the VMWatch process is also killed by subscribing to shutdown/termination signals in the main process, and killing the VMWatch based off process ID.

## Example Binary Execution
Example execution from integration testing 
`
SIGNAL_FOLDER=/var/log/azure/Microsoft.ManagedServices.ApplicationHealthLinux/events VERBOSE_LOG_FILE_FULL_PATH=/var/log/azure/Microsoft.ManagedServices.ApplicationHealthLinux/vmwatch.log ./var/lib/waagent/Extension/bin/VMWatch/vmwatch_linux_amd64 --config /var/lib/waagent/Extension/bin/VMWatch/vmwatch.conf --input-filter disk_io:outbound_connectivity
`

## Release/Packaging
In addition to the arm64 or amd64 VMWatch binaries, `vmwatch.conf` will be expected to be present in the bin/VMWatch directory for VMWatch process to read. VMWatch will also be populating and sharing eventsFolder with ApplicationHealth, so events can be viewed in Kusto. The verbose logs of VMWatch will be written to `vmwatch.log`.

### CODE REVIEW

File: misc/HandlerManifest.json [hunk 1]
```diff
@@ -9,5 +9,21 @@
     "rebootAfterInstall": false,
     "reportHeartbeat": false,
     "updateMode": "UpdateWithInstall"
+  },
+  "resourceLimits": {
+    "services": [
```
Comment (zmyzheng) [inline misc/HandlerManifest.json:14]: Are they running as backend services? If not, the service level resource restriction might not work.

Comment (zmyzheng) [inline misc/HandlerManifest.json:14]: Are they running as backend services? If not, the service level resource restriction might not work.

File: main/vmWatch.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	ctx.Log("event", fmt.Sprintf("Execute VMWatch %s", cmdToString(cmd)))
+
+	output, err := cmd.CombinedOutput()
+
+	pid := -1
+	if cmd.Process != nil {
+		pid = cmd.Process.Pid
+	}
+
+	defer func() {
```
Comment (zmyzheng) [inline main/vmWatch.go:59]: Just curious, why defer code block is needed here? If moving the content inside defer out of it, will the code still work?

**Review by zmyzheng — COMMENTED** [review]
Left some comments

File: main/vmWatch.go (hunk 2 above)
Comment (frank-pang-msft) [reply main/vmWatch.go:59]: Discussed on IM, defer will always run even on failure

File: misc/HandlerManifest.json (hunk 1 above)
Comment (frank-pang-msft) [reply misc/HandlerManifest.json:14]: Will remove

File: integration-test/test/6_vmwatch.bats [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    [[ "$output" == *'--config /var/lib/waagent/Extension/bin/vmwatch.conf'* ]]
+    [[ "$output" == *'--input-filter disk_io:outbound_connectivity'* ]]
+    [[ "$output" == *'Env: [SIGNAL_FOLDER=/var/log/azure/Microsoft.ManagedServices.ApplicationHealthLinux/events VERBOSE_LOG_FILE_FULL_PATH=/var/log/azure/Extension/vmwatch.log]'* ]]
+    [[ "$output" == *'VMWatch is running'* ]]
+    [[ "$output" == *'VMWatch process failed'* ]]
+
+    verify_substatus_item "$status_file" AppHealthStatus success "Application found to be healthy"
+    verify_substatus_item "$status_file" ApplicationHealthState transitioning Initializing
+    verify_substatus_item "$status_file" VMWatch error "VMWatch failed: .* Err: signal: terminated.*"
+}
```
Comment (klugorosado) [inline integration-test/test/6_vmwatch.bats:251]: Do you think the integration tests should also test the communication between VMWatch and AppHealth via HTTPS?

Comment (frank-pang-msft) [reply integration-test/test/6_vmwatch.bats]: VMWatch doesn't actually communicate with AppHealth via any protocol or networking. Today, AppHealth just executes VMWatch command and tracks whether the VMWatch process is running or not.

**Review by zmyzheng — APPROVED** [review]

**Review by zmyzheng — APPROVED** [review]

File: main/vmWatch.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	Status VMWatchStatus
+	Error  error
+}
+
+func (r *VMWatchResult) GetMessage() string {
+	switch r.Status {
+	case Disabled:
+		return "VMWatch is disabled"
+	case Failed:
+		return fmt.Sprintf("VMWatch failed: %s", r.Error.Error())
```
Comment (frank-pang-msft) [inline main/vmWatch.go:44]: handle nils

File: main/cmds.go [hunk 5]
```diff
@@ -121,6 +133,19 @@ func enable(ctx *log.Context, h vmextension.HandlerEnvironment, seqNum int) (str
 			return "", errTerminated
 		}
 
+		// If VMWatch was never supposed to run, it will be in Disabled state, so we do not need to read from the channel
+		// If VMWatch failed to execute, we will do not need to read from the channel
+		// Only if VMWatch is currently running do we need to check if it is terminated
+		if vmWatchResult.Status == Running {
+			select {
+			case vmWatchResult = <-vmWatchResultChannel:
+				ctx.Log("error", fmt.Sprintf("VMWatch process failed. %s", vmWatchResult.Error.Error()))
```
Comment (frank-pang-msft) [inline main/cmds.go:142]: handle nils

File: main/cmds.go [hunk 6]
```diff
@@ -121,6 +133,19 @@ func enable(ctx *log.Context, h vmextension.HandlerEnvironment, seqNum int) (str
...
+				close(vmWatchResultChannel)
+			default:
+				ctx.Log("event", "VMWatch is running")
```
Comment (frank-pang-msft) [inline main/cmds.go:145]: to avoid redundant logs, we should omit, or only log state changes

File: test.Dockerfile [hunk 7]
```diff
@@ -24,4 +24,4 @@ RUN ln -s /var/lib/waagent/fake-waagent /sbin/fake-waagent && \
 # Copy the handler files
 COPY misc/HandlerManifest.json ./Extension/
 COPY misc/applicationhealth-shim ./Extension/bin/
-COPY bin/applicationhealth-extension ./Extension/bin/
\ No newline at end of file
+COPY bin/applicationhealth-extension ./Extension/bin/
```
Comment (frank-pang-msft) [inline test.Dockerfile:27]: revert

File: main/cmds.go [hunk 8]
```diff
@@ -179,6 +204,9 @@ func enable(ctx *log.Context, h vmextension.HandlerEnvironment, seqNum int) (str
 			substatuses = append(substatuses, NewSubstatus(SubstatusKeyNameCustomMetrics, customMetricsStatusType, probeResponse.CustomMetrics))
 		}
 
+		// VMWatch will be built-in to GuestHealthFramework, so we will always display VMWatch as substatus
+		substatuses = append(substatuses, NewSubstatus(SubstatusKeyNameVMWatch, vmWatchResult.Status.GetStatusType(), vmWatchResult.GetMessage()))
```
Comment (frank-pang-msft) [inline main/cmds.go:208]: Update to only disable if VMWatchSettings is present

File: integration-test/env/README.md [hunk 9]
```diff
@@ -7,10 +7,13 @@ integration testing Docker image.
...
-    ├── bin/                    <-- docker image build pushes the extension binary here
-    ├── config/                 <-- tests push 0.settings file here
-    └── status/                 <-- extension should write here
+├── Extension/                  
+├   ├── HandlerManifest.json    <-- docker image build pushes it here
+├   ├── HandlerEnvironment.json <-- the extension reads this
+├   ├── bin/                    <-- docker image build pushes the extension binary here
+├   ├── config/                 <-- tests push 0.settings file here
+├   └── status/                 <-- extension should write here
+└── VMWatch/
```
Comment (frank-pang-msft) [inline integration-test/env/README.md:16]: update path

File: .gitignore [hunk 10]
```diff
@@ -12,6 +12,8 @@ applicationhealth-extension-linux
 _obj
 _test
 testbin
+integration-test/result
```
Comment (frank-pang-msft) [inline .gitignore:15]: remote integration-test/result

File: .github/workflows/merge_validation_v2.yml [hunk 11]
```diff
@@ -91,5 +92,13 @@ jobs:
...
 
     - name: Run Integration Tests
-      run: sudo bats integration-test/test
+      continue-on-error: true
+      run: |
+        mkdir -p integration-test/test/.bats/run-logs/
+        sudo bats integration-test/test
+      working-directory: ${{ env.repo_root }}
+
+    - name: Run Failing Integration Tests
```
Comment (frank-pang-msft) [inline .github/workflows/merge_validation_v2.yml:101]: Change name to "Retry Failing Integration Tests"

File: main/cmds.go [hunk 12]
```diff
@@ -121,6 +133,19 @@ func enable(ctx *log.Context, h vmextension.HandlerEnvironment, seqNum int) (str
...
 		}
 
+		// If VMWatch was never supposed to run, it will be in Disabled state, so we do not need to read from the channel
+		// If VMWatch failed to execute, we will do not need to read from the channel
+		// Only if VMWatch is currently running do we need to check if it is terminated
+		if vmWatchResult.Status == Running {
+			select {
+			case vmWatchResult = <-vmWatchResultChannel:
+				ctx.Log("error", fmt.Sprintf("VMWatch process failed. %s", vmWatchResult.Error.Error()))
+				close(vmWatchResultChannel)
```
Comment (frank-pang-msft) [inline main/cmds.go:143]: writer should be the one closing the channel, not the reader

File: main/schema.go [hunk 13]
```diff
@@ -11,57 +11,82 @@ import (
...
+        "minimum": 1,
+        "maximum": 24
+      },
+      "gracePeriod": {
+        "description": "Optional - The amount of time in seconds the application will default to 'Initializing' state if no valid health state is observed numberOfProbes consecutive times.",
+        "type": "integer",
+        "minimum": 5,
+        "maximum": 14400
+      },
+      "vmWatchSettings": {
```
Comment (frank-pang-msft) [inline main/schema.go:53]: Make vmWatchSettings default to nil if not specified.

enabled default=false
tests default=[ "disk_io", "outbound_connectivity", "clockskew", "az_storage_blob" ]
parameterOverrides default={}

File: main/handlersettings.go [hunk 14]
```diff
@@ -81,15 +85,22 @@ func (h handlerSettings) validate() error {
 	return nil
 }
 
+type vmWatchSettings struct {
```
Comment (frank-pang-msft) [inline main/handlersettings.go:88]: add VMWatchSettings UTs

File: main/vmWatch.go [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+	}
+}
+
+func executeVMWatch(ctx *log.Context, s vmWatchSettings, h vmextension.HandlerEnvironment, vmWatchResultChannel chan VMWatchResult) {
+	pid := -1
+	combinedOutput := &bytes.Buffer{}
+	var vmWatchErr error
+
+	// Best effort to start VMWatch process 3 times
+	for i := 1; i <= 3; i++ {
```
Comment (Hafthor) [inline main/vmWatch.go:56]: seems like this loop will always retry three times, even on success?

**Review by Hafthor — COMMENTED** [review]
just a first pass at this review -- might try to take another look at it later.

File: main/vmWatch.go (hunk 15 above)
Comment (frank-pang-msft) [reply main/vmWatch.go:56]: VMWatch process should run indefinitely, but if there are any failures that occur, we do best effort of 3 attempts to respin the process

File: main/vmWatch.go [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+		}
+
+		ctx.Log("event", fmt.Sprintf("Attempt %d: Setup VMWatch command: %s\nArgs: %v\nDir: %s\nEnv: %v\n", i, cmd.Path, cmd.Args, cmd.Dir, cmd.Env))
+
+		combinedOutput.Reset()
+		cmd.Stdout = combinedOutput
+		cmd.Stderr = combinedOutput
+
+		// Start command
+		err = cmd.Start()
```
[the remark this answers is not in the document]
Comment (frank-pang-msft) [reply main/vmWatch.go:72]: Yes we should is there a best practice to do that?

[the remark this answers is not in the document]
Comment (frank-pang-msft) [reply main/vmWatch.go:72]: Maybe by tracking cmd from main?

File: main/vmWatch.go [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+		return fmt.Sprintf("VMWatch failed: %s", r.Error.Error())
+	default:
+		return "VMWatch is running"
+	}
+}
+
+func executeVMWatch(ctx *log.Context, s vmWatchSettings, h vmextension.HandlerEnvironment, vmWatchResultChannel chan VMWatchResult) {
+	pid := -1
+	combinedOutput := &bytes.Buffer{}
+	var vmWatchErr error
```
[the remark this answers is not in the document]
Comment (frank-pang-msft) [reply main/vmWatch.go:53]: updated

File: main/vmWatch.go [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+
+		ctx.Log("event", fmt.Sprintf("Attempt %d: Setup VMWatch command: %s\nArgs: %v\nDir: %s\nEnv: %v\n", i, cmd.Path, cmd.Args, cmd.Dir, cmd.Env))
+
+		combinedOutput.Reset()
+		cmd.Stdout = combinedOutput
+		cmd.Stderr = combinedOutput
+
+		// Start command
+		err = cmd.Start()
+		if cmd.Process != nil {
```
[the remark this answers is not in the document]
Comment (frank-pang-msft) [reply main/vmWatch.go:73]: Updated to assign to -1 if Start failed. Lmk if that works

File: main/vmWatch.go [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+		// VMWatch should run indefinitely, if process exists we capture error
+		err = cmd.Wait()
+		if err != nil {
+			vmWatchErr = fmt.Errorf("[%v][PID %d] Err: %w\nOutput: %s", time.Now().UTC().Format(time.RFC3339), pid, err, combinedOutput.String())
+			ctx.Log("error", fmt.Sprintf("Attempt %d: VMWatch process exited: %s", i, vmWatchErr.Error()))
+			continue
+		}
+	}
+
+	defer func() {
```
[the remark this answers is not in the document]
Comment (frank-pang-msft) [reply main/vmWatch.go]: if we exit this method it means VMWatch (which should run indefinitely) has failed, so we always want to run defer method

File: main/vmWatch.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (frank-pang-msft) [reply main/vmWatch.go:81]: I have included in defer to cleanup the process if it exists

File: main/vmWatch.go [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+// We will setup and execute VMWatch as a separate process. Ideally VMWatch should run indefinitely,
+// but as a best effort we will attempt at most 3 times to run the process
+func executeVMWatch(ctx *log.Context, s *vmWatchSettings, h vmextension.HandlerEnvironment, vmWatchResultChannel chan VMWatchResult) {
+	var vmWatchErr error
+	defer func() {
+		if r := recover(); r != nil {
+			vmWatchErr = fmt.Errorf("%w\n Additonal Details: %+v", vmWatchErr, r)
+			ctx.Log("error", "Recovered %+v", r)
+		}
+		ctx.Log("error", fmt.Sprintf("Signaling VMWatchStatus is Failed due to reaching max of %d retries", VMWatchMaxProcessAttempts))
```
Comment (Hafthor) [inline main/vmWatch.go:59]: note that this defer func will happen whenever the executeVMWatch func exits... it's like a finally block... not a catch block. I think you want this error handling stuff to be conditionally run on if executeVMWatch does not reach the bottom of the func.. like perhaps if vmWatchErr != nil

File: main/vmWatch.go [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+
+func executeVMWatchHelper(ctx *log.Context, attempt int, vmWatchSettings *vmWatchSettings, handlerEnvironment vmextension.HandlerEnvironment) (err error) {
+	pid := -1
+	var cmd *exec.Cmd
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("Error: %w\n Additonal Details: %+v", err, r)
+			ctx.Log("error", "Recovered %+v", r)
+		}
+		killVMWatch(ctx, cmd)
```
Comment (Hafthor) [inline main/vmWatch.go:78]: I don't think you want to killVMWatch on successful exit of executeVMWatchHelper func.

File: main/vmWatch.go [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+		err = fmt.Errorf("[%v][PID -1] Attempt %d: VMWatch failed to start. Error: %w\nOutput: %s", time.Now().UTC().Format(time.RFC3339), attempt, err, combinedOutput.String())
+		ctx.Log("error", err.Error())
+		return err
+	}
+	pid = cmd.Process.Pid // cmd.Process should be populated on success
+	ctx.Log("event", fmt.Sprintf("Attempt %d: VMWatch process started with pid %d", attempt, pid))
+
+	// VMWatch should run indefinitely, if process exists we expect an error
+	err = cmd.Wait()
+	err = fmt.Errorf("[%v][PID %d] Attempt %d: VMWatch process exited. Error: %w\nOutput: %s", time.Now().UTC().Format(time.RFC3339), pid, attempt, err, combinedOutput.String())
```
Comment (Hafthor) [inline main/vmWatch.go:108]: I would think you would want to test the return of cmd.Wait() and only fmt.Errorf wrap it when err != nil.

File: main/vmWatch.go [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+	return filepath.Join(processDirectory, "VMWatch", binaryName)
+}
+
+func GetVMWatchEnvironmentVariables(parameterOverrides map[string]interface{}, h vmextension.HandlerEnvironment) []string {
+	var arr []string
+	for key, value := range parameterOverrides {
+		arr = append(arr, fmt.Sprintf("%s=%s", key, value))
+	}
+
+	arr = append(arr, fmt.Sprintf("SIGNAL_FOLDER=%s", HandlerEnvironmentEventsFolderPath))
```
Comment (zmyzheng) [inline main/vmWatch.go:178]: Can you change this to use h.HandlerEnvironment.EventsFolder?

**Pull request merged**